src/app/icd/ICDClientInfoPersistentStorage.h

```diff
@@ -0,0 +1,165 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <crypto/CHIPCryptoPAL.h>
+#include <crypto/SessionKeystore.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/StorageKeyName.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Delegate for providing storage key names and maximum key counter.
+ */
+class ICDStorageKeyDelegate
+{
+public:
+    /**
+     * Gets the storage key name for the specified index.
+     *
+     * @param aIndex The index of the key to retrieve.
+     *
+     * @return The storage key name.
+     */
+    virtual StorageKeyName GetKey(uint16_t aIndex) = 0;
+
+    /**
+     * @brief Gets the maximum key counter.
+     *
+     * @return The maximum key counter.
+     */
+    virtual uint16_t MaxKeyCounter() = 0;
+
+    virtual ~ICDStorageKeyDelegate(){};
+};
+
+/**
+ * The ICDClientInfoPersistentStorage class is an abstract interface that defines the operations
+ * for storing and retrieving ICD client information in persistent storage.
+ */
+class ICDClientInfoPersistentStorage
+{
+public:
+    /**
+     * Struct to store information about ICD Client Info.
+     */
+    struct ICDClientInfo
+    {
+        NodeId mPeerNodeId                 = kUndefinedNodeId;
+        uint32_t mStartICDCounter          = 0;
+        uint32_t mOffset                   = 0;
+        uint64_t mMonitoredSubject         = static_cast<uint64_t>(0);
+        Crypto::Aes128KeyHandle mSharedKey = Crypto::Aes128KeyHandle();
+
+        bool IsValid() { return mPeerNodeId != kUndefinedNodeId; }
+    };
+
+    /**
+     * Struct per identity representing persistent storage for ICD client information and key store
+     */
+    struct Storage
+    {
+        FabricIndex mFabricIndex                      = kUndefinedFabricIndex;
+        PersistentStorageDelegate * mpClientInfoStore = nullptr;
```

Comment:
We would have two high-level delegate, one is from @erjiaqing song's PR, which has ICDRegistrationDelegate, another is CheckInDelegate where it provides OnValidateCheckInMessage(System::PacketBufferHandle && payload) and OnReadyForSubscription or something similar. Both Delegate's implementation takes this ICDClientInfoPersistentStorageDelegate. If application don't wanna use ICDClientInfoPersistentStorageDelegate, and wanna implement the high-level API callbacks in CheckInDelegate and ICDRegistrationDelegate, it is fine.
Here, it seems it is not wise to put the decryption logic in ICDClientInfoPersistentStorageDelegateImpl(Take checkin message buffer, and return the clientInfo..)...

regarding "Note also that it may be desirable to just store all the other bits as metadata with the key, in a secure key store" , in our codebase, we are using SessionKeyStore, it seems we don't have actual SymmetricKeyStore, when reviewing the API SessionKeyStore, it seems we don't have way to store metaData....that is why I use PersistentStorage for metaData and KeyStore for raw key

Reply:
just refactor storage and clientInfo to dedicated header file, thanks